csharp/src/Google.Protobuf/Extension.cs

```diff
@@ -0,0 +1,109 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#endregion
+
+using System;
+
+namespace Google.Protobuf
+{
+    /// <summary>
+    /// Represents a non-generic extension definition
+    /// </summary>
+    public abstract class Extension
+    {
+        internal abstract Type TargetType { get; }
+
+        /// <summary>
+        /// Internal use. Creates a new extension with the specified field number.
+        /// </summary>
+        protected Extension(int number)
+        {
+            FieldNumber = number;
+        }
+
+        internal abstract IExtensionValue CreateValue();
+
+        /// <summary>
+        /// Gets the field number of this extension
+        /// </summary>
+        public int FieldNumber { get; }
+    }
+
+    /// <summary>
+    /// Represents a type-safe extension identifier used for getting and setting single extension values in <see cref="IExtensionMessage{T}"/> instances
+    /// </summary>
+    /// <typeparam name="TTarget">The message type this field applies to</typeparam>
+    /// <typeparam name="TValue">The field value type of this extension</typeparam>
+    public sealed class Extension<TTarget, TValue> : Extension where TTarget : IExtensionMessage<TTarget>
+    {
+        private readonly FieldCodec<TValue> codec;
+
+        /// <summary>
+        /// Creates a new extension identifier with the specified field number and codec
+        /// </summary>
+        public Extension(int number, FieldCodec<TValue> codec) : base(number)
+        {
+            this.codec = codec;
+        }
+
+        internal override Type TargetType => typeof(TTarget);
+
+        internal override IExtensionValue CreateValue()
+        {
+            return new ExtensionValue<TValue>(codec);
+        }
+    }
+
+    /// <summary>
+    /// Represents a type-safe extension identifier used for getting repeated extension values in <see cref="IExtensionMessage{T}"/> instances
+    /// </summary>
+    /// <typeparam name="TTarget">The message type this field applies to</typeparam>
+    /// <typeparam name="TValue">The repeated field value type of this extension</typeparam>
+    public sealed class RepeatedExtension<TTarget, TValue> : Extension where TTarget : IExtensionMessage<TTarget>
```

Comment:
I'm not sure if that'd be safe. I don't understand the full scope of your API but it sounds like it lacks the type safety that I had in mind with my implementation. 

I'd prefer to use the current implementation since it has a couple of advantages over extension identifiers in other language implementations; Specifically because the type itself is self documenting. 

You can't accidentally call the wrong method with it, you don't need to check the original proto file to see if it's repeated or not because you automatically know if it's repeated or not based on the type provided to the method.

If a consumer lacks type information (if you're using the base Extension type) we could add an IsRepeated method to Extension though it'd just be some simple runtime reflection.

```cs
public bool IsRepeated()
{
    return GetType().GetGenericTypeDefinition() == typeof(RepeatedField<,>);
}
```

Reply:
The usage of Extension is somehow like FieldDescriptor.  We do not separate into FieldDescriptor and RepeatedFieldDescriptor, thus better not to separate Extension as well. 

Plus C++ and Java  do not have RepeatedExtension, try best to keep different languages similar will benefit project that use multi languages for protobuf.
https://github.com/protocolbuffers/protobuf/blob/46a48e49aa8357bbeee8040819a35e59880e329a/java/core/src/main/java/com/google/protobuf/ExtensionLite.java#L48

Reply:
Right, but C++ and Java have language features that allow them to easily use one extension type safely, or they don't have their own APIs that can be reused effectively, *or* the extension APIs were designed before those features were created (they are old libraries after all). C# either doesn't have these features or has other APIs I could leverage for code reuse, making it easier to understand and maintain from a consumer side and developer side.
 
C# has FieldCodec and RepeatedField. If we want to expose a RepeatedField we have to use a FieldCodec to leverage the APIs in it to write, merge, and calculate the size of a field. FieldCodec is also capable of being used for single fields which means we can use it in single value fields. If we got rid of the RepeatedExtension type, I couldn't reuse FieldCodec and RepeatedField and the API would have to match C++ and Java which *doesn't match* other C# repeated field APIs.

Plus, docs for extensions in C++ and Java state

> (The exact implementation of extension identifiers is complicated and involves magical use of generics / templates – however, you don't need to worry about how extension identifiers work to use them.)

I'm still under the firm belief that having a separate type isn't going to matter that much since they're magic in other languages anyway and most consumers will exclusively be using them via the `IExtendableMessage` methods and the custom options accessors. So they probably won't realize they're separate types anyway.